### ISSUE TITLE
Where would I find documentation to be able to use this SDK?

### ISSUE THREAD

**Issue by malhayek2014** [LabTechSoftware/ConnectWisePSA-SDK#2]
I am trying to integrate my system with ConnectWise ticketing system. I would like to use this SDK to connect to get the job done, but I can't seem to find document that will help me.

where can I find documentation on how to use this SDK?

**Comment by tampadesign** [comment]
Sorry, but we do not have any documentation for this SDK. You will have to refer to some of the comments in the code but even there, your mileage may vary. 

This is a work-in-progress, unofficial SDK that we cannot reliably provide any support for at this time. If me or someone else on our web development team has extra time to assist you here, we will try.

As far as how to use the SDK, here's a quick few steps that may help you get started:

1) Download the SDK to your project (however way you wish) 

2) You will need to either make your own config file or edit the one in the SDK (not advisable -- may be overwritten when you update). The default config file is located in src/LabtechSoftware/ConnectwisePsaSdk/config/config.php. You can simply copy what's in this and store it somewhere in your project, then include it when needed. Regardless, you will need to fill in the blanks for the credentials array.

3a) If you are taking advantage of PHP's autoloading/namespace functionality, make sure the SDK is in a location where it can be found by the namespace resolver you have. Then simply make a new instance of the factory class:

```<?php
use LabtechSoftware\ConnectwisePsaSdk\ConnectwiseApiFactory;

$sdk = new ConnectwiseApiFactory();
```

3b) If you are not using namespaces/autoloading, you will need to include the factory class in your PHP code. Example: 

```<?php
include path/to/sdk/src/LabtechSoftware/ConnectwisePsaSdk/ConnectwiseApiFactory.php
$sdk = new ConnectwiseApiFactory();
```

4) Now that we have an instance of the "factory" class ready to go, we can use the Service Ticket library. This is also where we will pass in the config array/file as mentioned in step 2: 

```<?php
// Get the config array how you prefer, we will need to pass it as an argument to the make() method
// make() accepts either a path (string) to the config file or the config array itself
$config = 'path/to/your/config.php';

// Get a new instance of the Service Ticket class (src/LabtechSoftware/ConnectwisePsaSdk/ServiceTicket.php)
$serviceTicketApi = $sdk->make('ServiceTicket', $config);

// Now you can use any of the methods in the Service ticket class as needed
// For example, we can get 10 service tickets from ConnectWise:
$tickets = $serviceTicketApi->findServiceTickets(10);
```

Due to the lack of documentation, you will just have to look at the ServiceTicket.php class and refer to the code comments. We will try to add documentation when we can, but there is no timetable I can provide right now. 

Hope that helps

**Comment by malhayek2014** [comment]
Hello,
Thank you for your help. That was great information that should probably be in the read me file to get other going :)

I am able to connect to the API now. But I have some questions, I hoping you have that to help me keep moving.

When I use the `addOrUpdateServiceTicketViaCompanyId` method I will have to pass my companyID and array of the field that I need to update/insert. 
How should the array look like? Should the array's key be the field name of the database's table like below? If so how can I get a list of all the key list or database field name? lastly, is there a unique field that I can create for external application (to help me do the insert/update based on a unique value generated by an external application)?


array('Summary' => 'This is a new ticket for testing',
         'ProblemDescription' => 'Please ignore this ticket. it is for testing only');

**Comment by malhayek2014** [comment]
I looked over connectWise Documentation https://developer.connectwise.com/API_Documentation/Service_Ticket_API/AddOrUpdateServiceTicketViaCompanyIdentifier where i found a list of the fields "I think this is correct"

so I here is what I attempted
	$sdk = new LabtechSoftware\ConnectwisePsaSdk\ConnectwiseApiFactory();

	$serviceTicketApi = $sdk->make('ServiceTicket', $config);

	$b = array('companyId' => $companyID,
			   'serviceTicket' => array(
									'TicketNumber' => '',
									'Summary' => 'This is a test Ticket - Please Ignore',
									'SiteName' => '',
									'Board' => '',
									'ServiceType' => '',
									'Status' => 'N',
									'StatusName' => '',
									'AddressLine1' => '',
									'AddressLine2' => '',
									'City' => '',
									'StateId' => '',
									'Zip' => '',
									'Country' => '',
									'Resolution' => '',
									'RemoteInternalCompanyName' => 'Remote',
									'ProblemDescription' => 'This is the body of the ticket please ignore it. Created by Mike Via API',
									'SendingSrServiceRecid' => '',
									'DateReq' => '',
									'SubBillingMethodId' => '',
									'SubBillingAmount' => '',
									'SubDateAccepted' => '',
									'SubDateAcceptedUtc' => '',
									'CustupdateFlag' => '',
									'ContactName' => '',
									'ContactPhoneNumber' => '',
									'ContactPhoneExtension' => '',
									'ContactEmailAddress' => '',
									'NewTicketContactEmailLookup' => '',
									'XRefID' => '',
									'Location' => '',
									'BudgetHours' => '',
									'Priority' => '',
									'ProcessNotifications' => '',
									'SkipCallback' => '',
									'PoNumber' => '',
									'Source' => '',
									'OpportunityId' => '',
									'AgreementId' => '',
									'TeamId' => '',
									'Severity' => '',
									'Impact' => '',
									'Location' => '',
									'BudgetHours' => '',
									'Priority' => '',
									'ProcessNotifications' => '',
									'SkipCallback' => '',
									'PoNumber' => '',
									'Source' => '',
									'OpportunityId' => '',
									'AgreementId' => '',
									'TeamId' => '',
									'Severity' => '',
									'Impact' => '',
									'ServiceSubType' => '',
									'ServiceItem' => '',
									'LastUpdateDate' => '',
									'ActualHours' => '',
									'BoardID' => '',
									'EnteredDate' => '',
									'ClosedDate' => '',
									'CompanyRecId' => '',
									'ContactRecId' => '',
									'CompanyId' => '',
									'PortalTypeId' => '',
									'Approved' => '',
									'RequiredDate' => '',
									'DateReqUtc' => '',
									'ClosedFlag' => '',
								),
					'TicketConfiguration' => array(
												'Id' => '',
												'DeviceId' => ''
								),
					'TicketNote' => array(
										'ProcessNotifications' => '',
										'Id' => '',
										'NoteText' => '',
										'PortalIsInternalNote' => '',
										'PortalIsExternalNote' => '',
										'CustomerUpdatedFlag' => '',
										'IsPartOfDetailDescription' => '',
										'IsPartOfInternalAnalysis' => '',
										'IsPartOfResolution' => '',
										'MemberRecID' => '',
										'ContactRecID' => '',
										'DateCreated' => '',
										'CreatedBy' => '',
										'MemberFullName' => '',
										'DateCreatedUtc' => '',
									),
					'TicketTask' => array(
										'Id' => '',
										'SrServiceRecID' => '',
										'Notes' => '',
										'IsClosed' => '',
										'Priority' => '',
										'StartDate' => '',
										'Resolution' => '',
										'ChildRecID' => '',
										'ChildDesc' => '',
										'CustUpdateFlag' => ''
									),
					'DocumentInfo' => array(
										'Id' => '',
										'Title' => '',
										'FileName' => '',
										'ServerFileName' => '',
										'Path' => '',
										'LastUpdated' => '',
										'Owner' => '',
										'isLink' => '',
										'isImage' => '',
										'IsPublic' => '',
										'ChildDesc' => '',
										'Content' => ''
									)
				);
	
	$new_ticket = $serviceTicketApi->addOrUpdateServiceTicketViaCompanyId($companyID, $b);

	echo '<pre>';
	print_r($new_ticket);
	echo '</pre>';
	

But I get this error! perhaps I setup my array incorrectly.

`Fatal error: Uncaught exception 'LabtechSoftware\ConnectwisePsaSdk\ApiException' with message 'ConnectWise API Error: Encoding: object has no 'TicketNumber' property' in C:\phpsites\CW3\LabtechSoftware\ConnectwisePsaSdk\SoapApiRequester.php:29 Stack trace: #0 C:\phpsites\CW3\LabtechSoftware\ConnectwisePsaSdk\ServiceTicket.php(40): LabtechSoftware\ConnectwisePsaSdk\SoapApiRequester->makeRequest('AddOrUpdateServ...', Array) #1 C:\phpsites\CW3\index.php(170): LabtechSoftware\ConnectwisePsaSdk\ServiceTicket->addOrUpdateServiceTicketViaCompanyId('RDI', Array) #2 {main} thrown in C:\phpsites\CW3\LabtechSoftware\ConnectwisePsaSdk\SoapApiRequester.php on line 29`

**Comment by digitalfiz** [comment]
Set TicketNumber to 0 not ''. You might have to do that with any other properties that are numeric also. SoapClient acts funny when you give it invalid values.

**Comment by malhayek2014** [comment]
setting it to 0 did not solve the problem. I still get the same error. is my array setup correctly?

**Comment by digitalfiz** [comment]
I'm sorry I did not notice that. No you need to pass the service ticket to the second parameter.

Here I have modified your example.

```php

$serviceTicketApi = $sdk->make('ServiceTicket', $config);

$serviceTicket = array(
                                'TicketNumber' => '',
                                'Summary' => 'This is a test Ticket - Please Ignore',
                                'SiteName' => '',
                                'Board' => '',
                                'ServiceType' => '',
                                'Status' => 'N',
                                'StatusName' => '',
                                'AddressLine1' => '',
                                'AddressLine2' => '',
                                'City' => '',
                                'StateId' => '',
                                'Zip' => '',
                                'Country' => '',
                                'Resolution' => '',
                                'RemoteInternalCompanyName' => 'Remote',
                                'ProblemDescription' => 'This is the body of the ticket please ignore it. Created by Mike Via API',
                                'SendingSrServiceRecid' => '',
                                'DateReq' => '',
                                'SubBillingMethodId' => '',
                                'SubBillingAmount' => '',
                                'SubDateAccepted' => '',
                                'SubDateAcceptedUtc' => '',
                                'CustupdateFlag' => '',
                                'ContactName' => '',
                                'ContactPhoneNumber' => '',
                                'ContactPhoneExtension' => '',
                                'ContactEmailAddress' => '',
                                'NewTicketContactEmailLookup' => '',
                                'XRefID' => '',
                                'Location' => '',
                                'BudgetHours' => '',
                                'Priority' => '',
                                'ProcessNotifications' => '',
                                'SkipCallback' => '',
                                'PoNumber' => '',
                                'Source' => '',
                                'OpportunityId' => '',
                                'AgreementId' => '',
                                'TeamId' => '',
                                'Severity' => '',
                                'Impact' => '',
                                'Location' => '',
                                'BudgetHours' => '',
                                'Priority' => '',
                                'ProcessNotifications' => '',
                                'SkipCallback' => '',
                                'PoNumber' => '',
                                'Source' => '',
                                'OpportunityId' => '',
                                'AgreementId' => '',
                                'TeamId' => '',
                                'Severity' => '',
                                'Impact' => '',
                                'ServiceSubType' => '',
                                'ServiceItem' => '',
                                'LastUpdateDate' => '',
                                'ActualHours' => '',
                                'BoardID' => '',
                                'EnteredDate' => '',
                                'ClosedDate' => '',
                                'CompanyRecId' => '',
                                'ContactRecId' => '',
                                'CompanyId' => '',
                                'PortalTypeId' => '',
                                'Approved' => '',
                                'RequiredDate' => '',
                                'DateReqUtc' => '',
                                'ClosedFlag' => '',
                            ),
                'TicketConfiguration' => array(
                                            'Id' => '',
                                            'DeviceId' => ''
                            ),
                'TicketNote' => array(
                                    'ProcessNotifications' => '',
                                    'Id' => '',
                                    'NoteText' => '',
                                    'PortalIsInternalNote' => '',
                                    'PortalIsExternalNote' => '',
                                    'CustomerUpdatedFlag' => '',
                                    'IsPartOfDetailDescription' => '',
                                    'IsPartOfInternalAnalysis' => '',
                                    'IsPartOfResolution' => '',
                                    'MemberRecID' => '',
                                    'ContactRecID' => '',
                                    'DateCreated' => '',
                                    'CreatedBy' => '',
                                    'MemberFullName' => '',
                                    'DateCreatedUtc' => '',
                                ),
                'TicketTask' => array(
                                    'Id' => '',
                                    'SrServiceRecID' => '',
                                    'Notes' => '',
                                    'IsClosed' => '',
                                    'Priority' => '',
                                    'StartDate' => '',
                                    'Resolution' => '',
                                    'ChildRecID' => '',
                                    'ChildDesc' => '',
                                    'CustUpdateFlag' => ''
                                ),
                'DocumentInfo' => array(
                                    'Id' => '',
                                    'Title' => '',
                                    'FileName' => '',
                                    'ServerFileName' => '',
                                    'Path' => '',
                                    'LastUpdated' => '',
                                    'Owner' => '',
                                    'isLink' => '',
                                    'isImage' => '',
                                    'IsPublic' => '',
                                    'ChildDesc' => '',
                                    'Content' => ''
                                );

$new_ticket = $serviceTicketApi->addOrUpdateServiceTicketViaCompanyId($companyID, $serviceTicket);

echo '<pre>';
print_r($new_ticket);
echo '</pre>';
```

**Comment by malhayek2014** [comment]
I have changed anything ID to have 0 instead of ''. Now, I get a new error

Fatal error: Uncaught exception 'LabtechSoftware\ConnectwisePsaSdk\ApiException' with message 'ConnectWise API Error: System.Web.Services.Protocols.SoapException: Server was unable to read request. ---> System.InvalidOperationException: There is an error in XML document (2, 1223). ---> System.FormatException: The string '' is not a valid AllXsd value. at System.Xml.Schema.XsdDateTime..ctor(String text, XsdDateTimeFlags kinds) at System.Xml.XmlConvert.ToDateTime(String s, XmlDateTimeSerializationMode dateTimeOption) at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReader1.Read13_ServiceTicket(Boolean isNullable, Boolean checkType) at Microsoft.Xml.Serialization.GeneratedAssembly.XmlSerializationReader1.Read34_Item() at Microsoft.Xml.Serialization.GeneratedAssembly.ArrayOfObjectSerializer28.Deserialize(XmlSerializationReader reader) at System.Xml.Serialization.XmlSerializer.Deserialize(XmlReader xmlReader, String encodingStyle, XmlDeserializationEvents events) --- End of inner exc in C:\phpsites\CW3\LabtechSoftware\ConnectwisePsaSdk\SoapApiRequester.php on line 29



here is my Array

$b = array(	'TicketNumber' => 0,
					'Summary' => 'This is a test Ticket - Please Ignore',
					'SiteName' => '',
					'Board' => '',
					'ServiceType' => '',
					'Status' => 'N',
					'StatusName' => '',
					'AddressLine1' => '',
					'AddressLine2' => '',
					'City' => '',
					'StateId' => 'NV',
					'Zip' => '',
					'Country' => '',
					'Resolution' => '',
					'RemoteInternalCompanyName' => 'Remote',
					'ProblemDescription' => 'This is the body of the ticket please ignore it. Created by Mike Via API',
					'SendingSrServiceRecid' => 0,
					'DateReq' => '',
					'SubBillingMethodId' => 0,
					'SubBillingAmount' => '',
					'SubDateAccepted' => '',
					'SubDateAcceptedUtc' => '',
					'CustupdateFlag' => '',
					'ContactName' => '',
					'ContactPhoneNumber' => '',
					'ContactPhoneExtension' => '',
					'ContactEmailAddress' => '',
					'NewTicketContactEmailLookup' => '',
					'XRefID' => '',
					'Location' => '',
					'BudgetHours' => 0,
					'Priority' => '',
					'ProcessNotifications' => '',
					'SkipCallback' => '',
					'PoNumber' => '',
					'Source' => '',
					'OpportunityId' => 0,
					'AgreementId' => 0,
					'TeamId' => 0,
					'Severity' => '',
					'Impact' => '',
					'ServiceSubType' => '',
					'ServiceItem' => '',
					'LastUpdateDate' => '',
					'ActualHours' => '',
					'BoardID' => 0,
					'EnteredDate' => '',
					'ClosedDate' => '',
					'CompanyRecId' => 0,
					'ContactRecId' => 0,
					'CompanyId' =>  $companyID,
					'PortalTypeId' => 0,
					'Approved' => '',
					'RequiredDate' => '',
					'DateReqUtc' => '',
					'ClosedFlag' => '',
					'TicketConfiguration' => array(
												'Id' => 0,
												'DeviceId' => ''
					),
					'TicketNote' => array(
										'ProcessNotifications' => '',
										'Id' => 0,
										'NoteText' => '',
										'PortalIsInternalNote' => '',
										'PortalIsExternalNote' => '',
										'CustomerUpdatedFlag' => '',
										'IsPartOfDetailDescription' => '',
										'IsPartOfInternalAnalysis' => '',
										'IsPartOfResolution' => '',
										'MemberRecID' => 0,
										'ContactRecID' => 0,
										'DateCreated' => '',
										'CreatedBy' => '',
										'MemberFullName' => '',
										'DateCreatedUtc' => '',
					),
					'TicketTask' => array(
										'Id' => 0,
										'SrServiceRecID' => '',
										'Notes' => '',
										'IsClosed' => '',
										'Priority' => '',
										'StartDate' => '',
										'Resolution' => '',
										'ChildRecID' => 0,
										'ChildDesc' => '',
										'CustUpdateFlag' => ''
					),
					'DocumentInfo' => array(
										'Id' => 0,
										'Title' => '',
										'FileName' => '',
										'ServerFileName' => '',
										'Path' => '',
										'LastUpdated' => '',
										'Owner' => '',
										'isLink' => '',
										'isImage' => '',
										'IsPublic' => '',
										'ChildDesc' => '',
										'Content' => ''
					)
				);

**Comment by malhayek2014** [comment]
I finally got it to work. I had to switch all the true to "1" and all the false to "0". Also, I had to correct the datetime format from '2015-04-13 19:43:25' to '2015-04-13T19:43:25'.

Now, I am running into another issue when trying to insert into/Update a Service Ticket Note. I tried to use this method "updateTicketNote" as the ServiceTicket.php file advise, but I get this error

Fatal error: Uncaught exception 'LabtechSoftware\ConnectwisePsaSdk\ApiException' with message 'ConnectWise API Error: Function ("UpdateTicketNote") is not a valid method for this service' in C:\phpsites\CW3\LabtechSoftware\ConnectwisePsaSdk\SoapApiRequester.php:29 Stack trace: #0 C:\phpsites\CW3\LabtechSoftware\ConnectwisePsaSdk\ServiceTicket.php(346): LabtechSoftware\ConnectwisePsaSdk\SoapApiRequester->makeRequest('UpdateTicketNot...', Array) #1 C:\phpsites\CW3\send_new_tickets.php(250): LabtechSoftware\ConnectwisePsaSdk\ServiceTicket->updateTicketNote(Array, '0') #2 {main} thrown in C:\phpsites\CW3\LabtechSoftware\ConnectwisePsaSdk\SoapApiRequester.php on line 29

When I looked at the API documentation listed on connectWise.com, I did not see a method called "updateTicketNote". However I see a method called "AddOrUpdateTicketNote" but this method is only for version 2.0 of this API. Since I am using 1.5 I am not sure how can I insert/update comment to connect wise?

Thank you again for your help

**Comment by digitalfiz** [comment]
Can you pull the WSDL up for it?

You can get it like this:

http://domain.com/v4_6_release/apis/1.5/ServiceTicketAPI.asmx?wsdl

You can also see the documentation for the method here:
http://domain.com/v4_6_release/apis/1.5/ServiceTicketAPI.asmx?op=UpdateTicketNote

If that second link doesn't exist can you tell me the exact version of ConnectWise you are using? Maybe this method is in newer version. I see it in our version but we are perpetually on an alpha version. ConnectWise doesn't provide any information that I can find on when that method was added but I can definitely do some more digging.

**Comment by malhayek2014** [comment]
I have the WSDL file for it but I can't upload it here. for the second URL I get the following



ConnectWise Service Ticket API

The following operations are supported. For a formal definition, please review the Service Description.

    AddOrUpdateServiceTicketViaCompanyId
    Adds or updates a service ticket for a company identified by the text-based company id. If the service ticket number is 0, the service ticket is added. If non-zero, the existing service ticket with that ticket number is updated.

    AddOrUpdateServiceTicketViaManagedId
    Adds or updates a service ticket for a company identified by managed id. If the service ticket number is 0, the service ticket is added. If non-zero, the existing service ticket with that ticket number is updated.

    AddOrUpdateTicketProduct
    Add or Update a product on a ticket

    AddServiceTicketViaCompanyId
    Adds a new service ticket for a company identified by the text-based company id.

    AddServiceTicketViaManagedId
    Adds a new service ticket for a company identified by managed id.

    AddTicketProduct
    Add a product on a ticket

    DeleteServiceTicket
    Deletes a service ticket by the ticket number.

    DeleteTicketDocument
    Removes the document from the ticket

    DeleteTicketProduct
    Delete Product from a ticket

    FindServiceTickets
    Finds service ticket information by a set of conditions. If isOpen is set, it will pre-filter by the ticket status.

    GetServiceStatuses
    Gets the list of statuses available to the specified ticket.

    GetServiceTicket
    Gets a service ticket by the ticket number. If no service ticket exists with the given ticket number, null is returned.

    GetTicketCount
    Gets the count of service tickets that meet the specified conditions

    GetTicketDocuments
    Gets the documents attached to the specified ticket

    GetTicketProductList
    Get a list of products for the specified ticket

    LoadServiceTicket
    Gets a service ticket by the ticket number. If no service ticket exists with the given ticket number, an error is returned.

    SearchKnowledgebase
    Performs a Knowledgebase search using the specified parameters

    SearchKnowledgebaseCount
    Counts the Knowledgebase records that will be returned by performing the associated search.

    UpdateServiceTicketViaCompanyId
    Updates an existing service ticket for a company identified by the text-based company id.

    UpdateServiceTicketViaManagedId
    Updates an existing service ticket for a company identified by managed id.

    UpdateTicketProduct
    Update a product on a ticket

**Comment by digitalfiz** [comment]
Really interesting. Can you give me the version of ConnectWise your on? Here are the list of methods I get from that url:


AddOrUpdateServiceTicketViaCompanyId 
Adds or updates a service ticket for a company identified by the text-based company id. If the service ticket number is 0, the service ticket is added. If non-zero, the existing service ticket with that ticket number is updated.

AddOrUpdateServiceTicketViaManagedId 
Adds or updates a service ticket for a company identified by managed id. If the service ticket number is 0, the service ticket is added. If non-zero, the existing service ticket with that ticket number is updated.

AddOrUpdateTicketProduct 
Add or Update a product on a ticket

AddServiceTicketViaCompanyId 
Adds a new service ticket for a company identified by the text-based company id.

AddServiceTicketViaManagedId 
Adds a new service ticket for a company identified by managed id.

AddTicketProduct 
Add a product on a ticket

DeleteServiceTicket 
Deletes a service ticket by the ticket number.

DeleteTicketDocument 
Removes the document from the ticket

DeleteTicketProduct 
Delete Product from a ticket

FindServiceTickets 
Finds service ticket information by a set of conditions. If isOpen is set, it will pre-filter by the ticket status.

GetDocument 
Gets the document for the specified documentId

GetServiceStatuses 
Gets the list of statuses available to the specified ticket.

GetServiceTicket 
Gets a service ticket by the ticket number. If no service ticket exists with the given ticket number, null is returned.

GetTicketCount 
Gets the count of service tickets that meet the specified conditions

GetTicketDocuments 
Gets the documents attached to the specified ticket

GetTicketProductList 
Get a list of products for the specified ticket

LoadServiceTicket 
Gets a service ticket by the ticket number. If no service ticket exists with the given ticket number, an error is returned.

SearchKnowledgebase 
Performs a Knowledgebase search using the specified parameters

SearchKnowledgebaseCount 
Counts the Knowledgebase records that will be returned by performing the associated search.

UpdateServiceTicketViaCompanyId 
Updates an existing service ticket for a company identified by the text-based company id.

UpdateServiceTicketViaManagedId 
Updates an existing service ticket for a company identified by managed id.

UpdateTicketNote 
Add a new ticket note or update an existing ticket note by service ticket Recid.

UpdateTicketProduct 
Update a product on a ticket

**Comment by malhayek2014** [comment]
I just contacted CW tech support and here is what they said
"The cloud server works with by routing the API calls to specific versions of ConnectWise. The way it does this is by looking at the CompanyId for the request being sent. When viewing the WSDL in a browser you cannot specify the company. So you have to actually specify the version.

Here is an example
https://na.myconnectwise.net/v2015_3/apis/1.5/serviceticketapi.asmx?op=UpdateTicketNote

By listing the version it routes to the correct version available. When you make an actual SOAP call with XML it will automatically route to the correct version. So you will want to use v4_6_release for everything else."

I am not sure what version on CW we are on to be honest :( I requested that question from someone that is using CW in out company and I am waiting on an answer.


so my domain in the config array I am using this "domain"
I am not sure why the API is not working. it seems that the WSDL file is incorrect or it is pointing to the incorrect WSDL.

**Comment by malhayek2014** [comment]
I believe we are using v2015.2 (28258)

**Comment by digitalfiz** [comment]
I don't suppose upgrading is an option?

If not I think the only way to add notes is to update the whole ticket via UpdateServiceTicketViaCompanyId.

**Comment by malhayek2014** [comment]
Hum, I don't know if it is an option or not I will have to discuss that with the CW team. but what version should we be on for it to work as it should?

**Comment by digitalfiz** [comment]
I'm not really sure. If you look at the wsdl for your version it shows the
method exists. Maybe its a bug in the cloud routing?

https://na.myconnectwise.net/v2015_2/apis/1.5/serviceticketapi.asmx?op=UpdateTicketNote

And v2015.1 also:

https://na.myconnectwise.net/v2015_1/apis/1.5/serviceticketapi.asmx?op=UpdateTicketNote

**Comment by malhayek2014** [comment]
Any way I can figure out where is the cloud is routing the API to based on my companyID? this may help me figure out what it is using.

**Comment by digitalfiz** [comment]
Sorry I don't really know much about the cloud version of the CW api. We have never used it. You can probably recreate the call in plain PHP with SoapClient and send it too support to see if they can help you. I know they wont support our unoffical SDK but if you can show them a less abstracted example maybe they can help you track what the issue might be down.

**Comment by malhayek2014** [comment]
I will try to create the call using plain PHP I never used the SOAP extension so I will have to experiment with it.
I will let you know the outcome

**Comment by malhayek2014** [comment]
Hello,
How can I get a print out of the request just before making the soap call?
I am trying to debug something's and would like to see how my call looks like just before it is sent to the API?

Thank you

**Comment by digitalfiz** [comment]
Here is a stackover example of overriding the SoapClient class to make your own so you can do such a thing:

http://stackoverflow.com/a/1626572/1533130

You can do it without overriding the class but only after you make the request using `$client->__getLastRequest()`. 

But I think overriding would be best for your purposes.

**Comment by malhayek2014** [comment]
Hello- 
sorry I did not update you earlier as I have been busy trying to get the API to work. The CW support team came back and said I should use the 2.0 API not the 1.5. I have made modification to your API to work for the 2.0 API. Now I am using the 2.0 API but running into one issue where when I made the second request using the AddOrUpdateServiceTicketViaCompanyIdentifier to existing ticket the value of the BoardID seems to change for some reason that I can't explain. even though I send the second request with the correct BoardID. it seems to be a bug in the API.

I turned on the debugger in CW and I can see my request is being received correctly as I send it. But it seems that there is a blank request before each request I send which could be causing this issue. I am still investigating and I will update you once I have something useful.


Thank you
